### PR TITLE
Fix misinterpretation of Parquet's legacy ARRAY schemas.

### LINKED ISSUE #6967: [BUG] Parquet List corner cases fail to be parsed
**Describe the bug**
The tests `NO GROUP LIST`,  `SINGLE GROUP LIST`, and `SINGLE GROUP TUPLE LIST` as a part of https://github.com/NVIDIA/spark-rapids/pull/6914 fail because we are unable to understand these corner cases when working with the footer of the file.  The CUDF code appears to do the right thing, but in the plugin that we are throwing exceptions saying that it is not the proper format.

### PR DESCRIPTION
**Pull request by mythrocks**
Fixes #6967.

This change fixes how the `spark-plugin` interprets the legacy representations of ARRAY columns in Parquet.

The rules for interpreting ARRAY schemas are listed in the [Parquet `LogicalType` spec](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#backward-compatibility-rules).

Before this change, only the currently recommended ARRAY schema was supported: i.e. `(List)` columns must have a `REPEATED` group child, itself containing the element column named `element`. An example for `List<int32>` is listed below:
```
 # Schema
 <pyarrow._parquet.ParquetSchema object at 0x7f0b78f426c0>
required group field_id=-1 spark_schema {
  optional group field_id=-1 f (List) {
    repeated group field_id=-1 list {
      optional int32 field_id=-1 element;
    }
  }
}
```
The Parquet spec does extend support for legacy formats. For instance, the following is also a valid schema for `List<int32>`:
```
 # Schema
 <pyarrow._parquet.ParquetSchema object at 0x7f279fbb9600>
required group field_id=-1 spark_schema {
  required group field_id=-1 f (List) {
    repeated int32 field_id=-1 array;
  }
}
```
This commit allows for the the correct interpretation for the schema above, by adding support for Parquet's backward compatibility rules for ARRAY.

Depends on https://github.com/rapidsai/cudf/pull/13277.

### CODE REVIEW

**Comment by mythrocks** [comment]
Build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala [hunk 1]
```diff
@@ -809,6 +827,59 @@ private case class GpuParquetFileFilterHandler(@transient sqlConf: SQLConf) {
     }
   }
 
+  // scalastyle:off
+  // Implement Parquet LIST backwards-compatibility rules.
+  // See: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#backward-compatibility-rules
+  // scalastyle:on
+  // Inspired by Apache Spark's ParquetSchemaConverter.isElementType().
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala:834]: add a permalink to the "inspiring" code

Comment (mythrocks) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetScan.scala]: Agreed. I've added a link to the inspiration code.

**Review by revans2 — DISMISSED** [review]

**Comment by mythrocks** [comment]
Note that this doesn't (completely) address #6968 or #6970. Those weren't the immediate concern.
I'll chase this down while rapidsai/cudf/pull/13277 negotiates its way to `spark-rapids-jni`. I'll include the fix here if I can find it in time.

**Review by gerashegalov — APPROVED** [review]

**Review by gerashegalov — DISMISSED** [review]

**Comment by mythrocks** [comment]
Build

**Comment by mythrocks** [comment]
The CI failure is because the fix in rapidsai/cudf#13277 isn't available in the published `spark-rapids-jni` dependency. I will retry this once it's available.

**Comment by mythrocks** [comment]
Build

**Comment by mythrocks** [comment]
I've kicked off a new build. A new `spark-rapids-jni` artifact was published about half an hour ago.

**Comment by mythrocks** [comment]
The build looks like it's failing `verify` because of a missing entry for `filecache.checkStale` in `configs.md`.

Depends on https://github.com/NVIDIA/spark-rapids/pull/8241.

**Comment by mythrocks** [comment]
I have cherry-picked the changes from #8241 to get past the CI failures.

**Review by revans2 — CHANGES_REQUESTED** [review]
We cannot pull in someone else's PR to fix your build issues.  All you needed to do was to update the docs on your original PR.

**Comment by mythrocks** [comment]
> We cannot pull in someone else's PR to fix your build issues.  All you needed to do was to update the docs on your original PR.

Hmm. I wasn't sure what was to be done. The documentation change is unrelated to this PR. The change would also have been identical to the change in #8241. 

I figured if #8241 is merged first, this cherry-pick would have been akin to my rebasing this PR. 

I'm happy to revert and redo.

**Review by revans2 — APPROVED** [review]

File: docs/configs.md [hunk 2]
```diff
@@ -42,6 +42,7 @@ Name | Description | Default Value | Applicable at
...
 <a name="cloudSchemes"></a>spark.rapids.cloudSchemes|Comma separated list of additional URI schemes that are to be considered cloud based filesystems. Schemes already included: abfs, abfss, dbfs, gs, s3, s3a, s3n, wasbs. Cloud based stores generally would be total separate from the executors and likely have a higher I/O read cost. Many times the cloud filesystems also get better throughput when you have multiple readers in parallel. This is used with spark.rapids.sql.format.parquet.reader.type|None|Runtime
+<a name="filecache.checkStale"></a>spark.rapids.filecache.checkStale|Controls whether the cached is checked for being out of date with respect to the input file. When enabled, the data that has been cached locally for a file will be invalidated if the file is updated after being cached. This feature is only necessary if an input file for a Spark application can be changed during the lifetime of the application. If an individual input file will not be overwritten during the Spark application then performance may be improved by setting this to false.|false|Startup
```
Comment (gerashegalov) [inline docs/configs.md:45]: Should we leave it to chance how GH will deal with the upmerge?

Comment (mythrocks) [reply docs/configs.md:45]: I was afraid of this. (It's why I pulled that change into this branch before.)

I'm checking this locally.

Comment (mythrocks) [reply docs/configs.md:45]: Nope, no problem here. The change is identical, so the merge is clean. This should be safe to go in.

**Review by gerashegalov — APPROVED** [review]

**Comment by mythrocks** [comment]
Thanks for the reviews, chaps. This is merged now. I'll keep an eye out for any garbling in `configs.md`.

**Pull request merged**